signature/provider.go

```diff
@@ -0,0 +1,176 @@
+package signature
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/notaryproject/notation-core-go/signature"
+	"github.com/notaryproject/notation-go/plugin"
+)
+
+// builtInPluginMetaData is the builtin metadata used by builtinProvider.
+var builtInPluginMetaData = plugin.Metadata{
+	SupportedContractVersions: []string{plugin.ContractVersion},
+	Capabilities:              []plugin.Capability{plugin.CapabilitySignatureGenerator},
+	Name:                      "built-in",
+	Description:               "Notation built-in signer",
+	Version:                   plugin.ContractVersion,
+	URL:                       "https://github.com/notaryproject/notation-go",
+}
+
+// provider wraps a plugin.Runner and a signature.Signer.
+type provider interface {
+	plugin.Runner
+	signature.Signer
+	SetConfig(map[string]string)
+}
+
+// builtinProvider is a builtin provider implementation
+// which wraps the signature.Signature to support builtin signing method.
+// It only supports describe key and metadata command.
+type builtinProvider struct {
+	signature.LocalSigner
+}
+
+// newBuiltinProvider creates a builtinProvider to support local signing.
+func newBuiltinProvider(key crypto.PrivateKey, certChain []*x509.Certificate) (provider, error) {
+	builtinSigner, err := signature.NewLocalSigner(certChain, key)
+	if err != nil {
+		return nil, err
+	}
+	return &builtinProvider{
+		builtinSigner,
+	}, nil
+}
+
+// metadata provides metadata for builtinProvider.
+func (*builtinProvider) metadata() *plugin.Metadata {
+	// The only properties that are really relevant
+	// are the supported contract version and the capabilities.
+	// All other are just filled with meaningful data.
+	return &builtInPluginMetaData
+}
+
+// SetConfig sets config when signing.
+func (*builtinProvider) SetConfig(map[string]string) {}
+
+// Run implements the plugin workflow.
+//
+// builtinProvider only supports metadata and describe key.
+func (p *builtinProvider) Run(_ context.Context, req plugin.Request) (interface{}, error) {
+	switch req.Command() {
+	case plugin.CommandGetMetadata:
+		return p.metadata(), nil
+	case plugin.CommandDescribeKey:
+		req1 := req.(*plugin.DescribeKeyRequest)
+		return &plugin.DescribeKeyResponse{
+			KeyID: req1.KeyID,
+		}, nil
+	}
+	return nil, plugin.RequestError{
+		Code: plugin.ErrorCodeGeneric,
+		Err:  fmt.Errorf("command %q is not supported", req.Command()),
+	}
+}
+
+// externalProvider is an external provider implementation which will interact with plugin.
+// It supports all plugin commands.
+//
+// The detail implementation depends on the underlying plugin.
+//
+// It wraps a signature.Signature to support external signing.
```

Comment:
does it wrap a `signature.Signature`?

Reply:
Typo, I will fix it. It implements `Signer` interface.